src/MongoDB.Driver.Core/Core/Operations/EndTransactionOperation.cs

```diff
@@ -31,9 +31,21 @@ public abstract class EndTransactionOperation : IReadOperation<BsonDocument>
     {
         // private fields
         private MessageEncoderSettings _messageEncoderSettings;
+        private readonly BsonDocument _recoveryToken;
         private readonly WriteConcern _writeConcern;
 
         // protected constructors
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EndTransactionOperation"/> class.
+        /// </summary>
+        /// <param name="recoveryToken">The recovery token.</param>
+        /// <param name="writeConcern">The write concern.</param>
+        protected EndTransactionOperation(BsonDocument recoveryToken, WriteConcern writeConcern)
+            : this(writeConcern)
+        {
+            _recoveryToken = recoveryToken;
+        }
+
```

Comment:
I don't like to see constructor initialization split across two constructors, each one responsible for only part of the initialization. The new constructor should look like this:

    protected EndTransactionOperation(BsonDocument recoveryToken, WriteConcern writeConcern)
    {
        _recoveryToken = recoveryToken;
        _writeConcern = Ensure.IsNotNull(writeConcern, nameof(writeConcern));
    }

I'm assuming the only reason you kept the original constructor is backward compatibility, even though this class is in Core and I'm not sure how important backward compatibility is in Core. In any case, the original constructor should delegate to the new constructor:

    protected EndTransactionOperation(WriteConcern writeConcern)
        : this(recoveryToken: null, writeConcern)
    {
    }

Reply:
Fair enough! I kept Core API compatibility in accordance to how we've handled similar cases in the past. I agree it'd be nice if we could just use the new constructor instead. I've taken your suggestions!